refactor(hardware-testing): Provide testing coverage for robots manufactured with no cameras

CaseyBatten opened the pull request:
# Overview
Previously the factory QC process involved a peripherals test for the camera during the robot assembly touchpoint. This adds a new `sku` argument to the script that gets called during this phase of QC. The `sku` argument indicates that during the test the tester will be asked to enter the SKU code.  This SKU will be checked to determine if we should not find a camera attached as `/dev/video2` at the factory.

What we do under this circumstance is we write a PASS or FAIL to the 'camera-image' test results of the script output CSV file.

The command to run the camera-removed version of the robot assembly qc is:
`python3 -m hardware_testing.production_qc.robot_assembly_qc_ot3 --sku`
The tester will be prompted to enter the SKU code/scan a barcode of the SKU
Enter the following SKU: `999-00279`
*NOTE*: This SKU will be written to EEPROM

## Test Plan and Hands on Testing

- [x] Run this test on a Flex without a camera and get the expected PASS
- [x] Run this test on a Flex with a camera and get the expected FAIL
- [x] Run the original test (leave out the `--sku` argument) on a Flex with a camera and get the expected PASS
- [x] Run the original test (leave out the `--sku` argument) on a Flex WITHOUT a camera and get the expected FAIL

## Changelog


## Review requests
Does this look like what we want this touchpoint to do now?

## Risk assessment
Mid/Low - affects factory SOP

CaseyBatten committed:
initial implementation of camera_removed arg and behavior

CaseyBatten committed:
add test results to CSV output

CaseyBatten committed:
remove changes from other PR

CaseyBatten commented:
Note: This PR is going to be updated to reflect the SKU storage method of HW validation

CaseyBatten committed:
transition to SKU version camera tests

CaseyBatten committed:
type ignores

CaseyBatten committed:
linting

CaseyBatten committed:
base branch change mistake

CaseyBatten committed:
more linting

vegano1 reviewed (approved):
nice, thank you

SyntaxColoring commented on hardware/opentrons_hardware/drivers/eeprom/types.py:
```diff
@@ -3,7 +3,7 @@
 from dataclasses import dataclass
 from enum import Enum
-from typing import Any, Optional
+from typing import Any, Optional, Set, Tuple
```
For future reference, we don't need to import these anymore in our version of Python. We can just do `set[tuple[PropId, Any]]`.

SyntaxColoring commented on hardware/opentrons_hardware/drivers/eeprom/types.py:
```diff
@@ -69,3 +71,14 @@ class EEPROMData:
     unit_number: Optional[int] = None
+    sku: Optional[str] = None
+
+    def to_set(self) -> Set[Tuple[PropId, Any]]:
```
How come this second tuple value is typed as `Any`? It looks like it could be something like:

```python
set[tuple[PropId, str | int]]
```

CaseyBatten replied:
The thinking was that for now the eeprom writer is only concerned with storing the serial and the sku alongside our format version but in the future there might be other things we want to store, but this can certainly be made more explicit.

SyntaxColoring commented on hardware-testing/hardware_testing/production_qc/robot_assembly_qc_ot3/test_peripherals.py:
```diff
@@ -244,12 +250,34 @@ def _get_user_confirmation(question: str) -> bool:
+            removed_result = CSVResult.PASS
+        except Exception as e:
+            print(
+                f"Confirming camera not attached failed with the following error: {e}"
```
As a general Python rule, this kind of thing is an antipattern:

```python
print(f"Something went wrong: {e}")
```

The problem is that `{e}` just does `str(e)`, and in general, `str(e)` is not guaranteed to make any sense. For example, try running this:

```python
try:
    d = {"a": 123}
    d["b"]  # oops
except Exception as e:
    print(f"Something went wrong: {e}")
```

This will print:

> Something went wrong: b

You can imagine how that would be confusing if the error happens somewhere deep and that's the only message you get out of the program.[^1] You'd _at least_ want it to say `KeyError`, and ideally also include a stack trace.

In this case, where you just want to report that an error happened but let the program continue, you should either:

* Use one of the functions from the (unfortunately messy) [`traceback` module](https://docs.python.org/3/library/traceback.html) to stringify *full* exception details, or at least enough of the details that they're guaranteed to make some sense.
* Use `log.exception(message)`, `log.error(message, exc_info=True)`, or `log.warning(message, exc_info=True)`. These will automatically include full exception details.

[^1]: Rarely, it can be okay if you know exactly what exception type you're dealing with and you can guarantee that `str(e)` is going to be OK. Like, if you're doing something like `except json.JSONDecodeError as e`, not `except Exception as e`.

SyntaxColoring reviewed (commented):
I'm not familiar with the actual substance of this code, so I don't have anything helpful to say about that, but here are some general Python nitpicks.

CaseyBatten committed:
set and tuple redundancy cleanup

SyntaxColoring reviewed (approved):
LGTM from my perspective, no reason for the exception printing thing to block merging. Thanks!

CaseyBatten committed:
ensure we use video2 to support older build compatibility

Older builds did not include the udev rule that added in the ot_system_camera so factory tags could be missing it

Andiiiiiiyy committed:
add camera active result

CaseyBatten committed:
consolidate the eeprom interfacing down to hardware-testing

CaseyBatten committed:
remove eeprom changes from hardware testing branch

CaseyBatten committed:
small fix

CaseyBatten committed:
linting

CaseyBatten committed:
other linting

CaseyBatten committed:
removal of unused import
